Add TypeScript definition

Artoria2e5 opened the pull request:
This definition covers the case with objects only as if it's an identity function, which should be its intended use-case. For anything else there should at least be something that strips the input of its callability, but I don't think Omit<T, '()'> is a thing. (Yes, they do understand `new`, but as far as I know there is no formal interface for describing what it does.)

Artoria2e5 committed:
Add simply typescript definition

This definition covers the case with objects only as if it's an identity
function, which should be its intended use-case. For anything else there
should at least be something that strips the input of its [[Call]] /
[[Construct]] ability, but I don't think `Omit<T, '()'>` is a
thing.

Artoria2e5 commented:
That... can't be my fault.

sindresorhus commented:
Thanks for creating a TypeScript definition. 🙌 Can you follow [this styleguide](https://github.com/sindresorhus/typescript-definition-style-guide)?

Some examples:
- https://github.com/sindresorhus/semver-regex/commit/039944bb5b3876913171b895488874b5c200e458
- https://github.com/sindresorhus/global-dirs/commit/0aae4d64887f77af94c4fb7ca12aa9f93c0a8291
- https://github.com/sindresorhus/md5-hex/commit/bbbfeff840f26973fbffc16d4863ae6d120079f3

Artoria2e5 commented:
I don't like some of its choices but I will do it.

Artoria2e5 committed:
Add TypeScript tests

`require('tsd').default()` returns a Promise<Diagnostics[]>. Ava 0.0.4
cannot handle async callbacks or async addition of tests. Newer ava
mandates a single command which limits our ability to use node flags.
Doing execFileSync is possible, but hardcoding node_modules/.bin/tsd
feels wompy.

Well, let's forget about ava and just add an `&&` to the test script.

(Removed `xo` because nobody is using it and nothing passes it.)

sindresorhus committed:
Update package.json

sindresorhus committed:
Update package.json

sindresorhus committed:
Update index.test-d.ts

sindresorhus committed:
Update index.d.ts

sindresorhus committed:
Update index.test-d.ts
